File: app/src/main/java/com/backbase/mobileAssignment/utils/search/SearchUtils.java
package com.backbase.mobileAssignment.utils.search;

import java.util.List;

public class SearchUtils {
    private static SearchUtils searchUtils;
    private final IDataStructure iDataStructure;

    private SearchUtils(IDataStructure iDataStructure) {
        this.iDataStructure = iDataStructure;
    }

    public static SearchUtils getInstance(IDataStructure iDataStructure) {
        if (searchUtils == null) {
            searchUtils = new SearchUtils(iDataStructure);
        }
        return searchUtils;
    }


    public void insert(String word) {
        if(word == null)
            return;
        iDataStructure.insert(word.toLowerCase());
    }

    public boolean searchWord(String word) {
        if(word == null)
            return false;
        return iDataStructure.search(word.toLowerCase());
    }

    public List<String> getSuggestions(String prefix) {
        return iDataStructure.getSuggestions(prefix);
    }

}

File: app/src/main/java/com/backbase/mobileAssignment/utils/search/SearchUtils.kt
package com.backbase.mobileAssignment.utils.search

import com.backbase.mobileAssignment.data.database.entity.City

class SearchUtils private constructor(private val iDataStructure: IDataStructure) {
    fun insert(city: City?) {
        iDataStructure.insert(city)
    }

    fun searchWord(city: City?): Boolean {
        return iDataStructure.search(city)
    }

    fun getSuggestions(prefix: String?): List<City?> {
        return iDataStructure.getSuggestions(prefix)
    }

    fun fetchAll(): List<City?> {
        return iDataStructure.fetchAll()
    }

    companion object {
        private var searchUtils: SearchUtils? = null
        fun getInstance(iDataStructure: IDataStructure): SearchUtils {
            if (searchUtils == null) {
                searchUtils = SearchUtils(iDataStructure)
            }
            return searchUtils!!
        }
    }
}